### PR TITLE
Reconfigure how Protocol Specs are created WRT EVM condiguration

### PR DESCRIPTION
**Pull request by shemnon**
## PR description

Make the max code size and max initcode size a part of the EVM configuration. As part of the change we need to move the tasks CodeFactory once handled as a static class and move it into the EVM. This has a nice follow on effect that we don't need to pass in max EOF versions or max code sizes anymore.

This is also required to address a recent EOF change limiting the total size of containers.

### Thanks for sending a pull request! Have you done the following?

- [ ] Checked out our [contribution guidelines](https://github.com/hyperledger/besu/blob/main/CONTRIBUTING.md)?
- [ ] Considered documentation and added the `doc-change-required` label to this PR [if updates are required](https://wiki.hyperledger.org/display/BESU/Documentation).
- [ ] Considered the changelog and included an [update if required](https://wiki.hyperledger.org/display/BESU/Changelog).
- [ ] For database changes (e.g. KeyValueSegmentIdentifier) considered compatibility and performed forwards and backwards compatibility tests

### Locally, you can run these tests to catch failures early:

- [ ] unit tests: `./gradlew build`
- [ ] acceptance tests: `./gradlew acceptanceTest`
- [ ] integration tests: `./gradlew integrationTest`
- [ ] reference tests: `./gradlew ethereum:referenceTests:referenceTests`

### CODE REVIEW

File: evm/src/main/java/org/hyperledger/besu/evm/code/CodeV1Validation.java [hunk 1]
```diff
@@ -67,20 +75,25 @@ private CodeV1Validation() {
    * @param layout The parsed EOFLayout of the code
```
Comment (jflo) [inline evm/src/main/java/org/hyperledger/besu/evm/code/CodeV1Validation.java:75]:
```suggestion
   * @param layout The parsed EOFLayout of the code to validate both the bytecode and stack of
```

**Review by jflo — APPROVED** [review]
I have some suggestions to improve readability, but otherwise this is a broad improvement.

**Comment by shemnon** [comment]
Moving EOF aspects to #7250, this PR remains to discuss how to handle protocol spec changes WRT EVMConfiguration.

File: ethereum/eth/src/test/java/org/hyperledger/besu/ethereum/eth/sync/fastsync/PivotBlockRetrieverTest.java [hunk 2]
```diff
@@ -269,6 +271,7 @@ public void shouldQueryBestPeersFirst(final DataStorageFormat storageFormat) {
 
   @ParameterizedTest
   @ArgumentsSource(PivotBlockRetrieverTestArguments.class)
+  @DisabledOnOs(OS.MAC)
```
Comment (jflo) [inline ethereum/eth/src/test/java/org/hyperledger/besu/ethereum/eth/sync/fastsync/PivotBlockRetrieverTest.java:274]: You can drop this once this pr merges: https://github.com/hyperledger/besu/pull/7253

File: evm/src/main/java/org/hyperledger/besu/evm/code/CodeV1Validation.java [hunk 3]
```diff
@@ -107,13 +120,14 @@ public static String validate(final EOFLayout layout) {
    * @param eofLayout The EOF Layout
    * @return validation code, null otherwise.
    */
-  public static String validateCode(final EOFLayout eofLayout) {
```
Comment (jflo) [inline evm/src/main/java/org/hyperledger/besu/evm/code/CodeV1Validation.java:110]: love seeing less statics

**Review by jflo — APPROVED** [review]

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/ProtocolScheduleBuilder.java [hunk 4]
```diff
@@ -118,18 +117,18 @@ private void initSchedule(
     final MainnetProtocolSpecFactory specFactory =
         new MainnetProtocolSpecFactory(
             chainId,
-            config.getContractSizeLimit(),
-            config.getEvmStackSize(),
             isRevertReasonEnabled,
             config.getEcip1017EraRounds(),
-            evmConfiguration,
+            evmConfiguration.overrides(
+                config.getContractSizeLimit(), OptionalInt.empty(), config.getEvmStackSize()),
```
[the remark this answers is not in the document]
Comment (shemnon) [reply ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/ProtocolScheduleBuilder.java:123]: GenesisConfigOptions currently does not do anything but report direct fields.  Packaging up config objects is handled by the ProtocolSpecBuilder.  We may want to factor out how fee markets are built (for enterprise) and the system contract addresses into providers or a new record (for system contracts). Sounds like a different PR TBH.

File: ethereum/eth/src/test/java/org/hyperledger/besu/ethereum/eth/sync/fastsync/PivotBlockRetrieverTest.java (hunk 2 above)
Comment (shemnon) [reply ethereum/eth/src/test/java/org/hyperledger/besu/ethereum/eth/sync/fastsync/PivotBlockRetrieverTest.java]: remved.

File: evm/src/main/java/org/hyperledger/besu/evm/code/CodeV1Validation.java (hunk 1 above)
Comment (shemnon) [reply evm/src/main/java/org/hyperledger/besu/evm/code/CodeV1Validation.java]: dome slightly different

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/MainnetProtocolSpecs.java [hunk 5]
```diff
@@ -107,30 +102,23 @@ public abstract class MainnetProtocolSpecs {
...
 
-  public static ProtocolSpecBuilder frontierDefinition(
-      final OptionalInt configContractSizeLimit,
-      final OptionalInt configStackSizeLimit,
-      final EvmConfiguration evmConfiguration) {
-    final int contractSizeLimit = configContractSizeLimit.orElse(FRONTIER_CONTRACT_SIZE_LIMIT);
-    final int stackSizeLimit = configStackSizeLimit.orElse(MessageFrame.DEFAULT_MAX_STACK_SIZE);
+  public static ProtocolSpecBuilder frontierDefinition(final EvmConfiguration evmConfiguration) {
+    final int stackSizeLimit =
+        evmConfiguration.evmStackSizeOverride().orElse(MessageFrame.DEFAULT_MAX_STACK_SIZE);
```
[the remark this answers is not in the document]
Comment (shemnon) [reply ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/MainnetProtocolSpecs.java]: Done.

**Pull request merged**